partiql-types/src/main/kotlin/org/partiql/value/PartiQLValue.kt

```diff
@@ -405,398 +446,115 @@ public abstract class SexpValue<T : PartiQLValue> : CollectionValue<T> {
-}
-
-@PartiQLValueExperimental
-public abstract class NullableTimeValue : NullableScalarValue<Time> {
-
-    override val type: PartiQLValueType = PartiQLValueType.NULLABLE_TIME
-
-    abstract override fun copy(annotations: Annotations): NullableTimeValue
-
-    abstract override fun withAnnotations(annotations: Annotations): NullableTimeValue
-
-    abstract override fun withoutAnnotations(): NullableTimeValue
-}
-
-@PartiQLValueExperimental
-public abstract class NullableTimestampValue : NullableScalarValue<Timestamp> {
-
-    override val type: PartiQLValueType = PartiQLValueType.NULLABLE_TIMESTAMP
-
-    abstract override fun copy(annotations: Annotations): NullableTimestampValue
-
-    abstract override fun withAnnotations(annotations: Annotations): NullableTimestampValue
-
-    abstract override fun withoutAnnotations(): NullableTimestampValue
-}
-
-@PartiQLValueExperimental
-public abstract class NullableIntervalValue : NullableScalarValue<Long> {
-
-    override val type: PartiQLValueType = PartiQLValueType.NULLABLE_INTERVAL
+    override val type: PartiQLValueType = PartiQLValueType.STRUCT
 
-    abstract override fun copy(annotations: Annotations): NullableIntervalValue
+    public abstract val fields: Sequence<Pair<String, T>>?
 
-    abstract override fun withAnnotations(annotations: Annotations): NullableIntervalValue
+    override val isNull: Boolean
+        get() = fields == null
 
-    abstract override fun withoutAnnotations(): NullableIntervalValue
-}
+    override fun iterator(): Iterator<Pair<String, T>> = fields!!.iterator()
 
-@PartiQLValueExperimental
-public abstract class NullableBagValue<T : PartiQLValue> : NullableCollectionValue<T> {
+    public abstract operator fun get(key: String): T?
 
-    override val type: PartiQLValueType = PartiQLValueType.NULLABLE_BAG
+    public abstract fun getAll(key: String): Iterable<T>
 
-    abstract override fun promote(): BagValue<T>
+    abstract override fun copy(annotations: Annotations): StructValue<T>
 
-    abstract override fun copy(annotations: Annotations): NullableBagValue<T>
+    abstract override fun withAnnotations(annotations: Annotations): StructValue<T>
 
-    abstract override fun withAnnotations(annotations: Annotations): NullableBagValue<T>
+    abstract override fun withoutAnnotations(): StructValue<T>
 
-    abstract override fun withoutAnnotations(): NullableBagValue<T>
+    /**
+     * See equality of IonElement StructElementImpl
+     *
+     * https://github.com/amazon-ion/ion-element-kotlin/blob/master/src/com/amazon/ionelement/impl/StructElementImpl.kt
+     *
+     * @param other
+     * @return
+     */
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is StructValue<*>) return false
+        if (other.annotations != this.annotations) return false
+
+        // both null.struct
+        if (other.isNull && this.isNull) return true
```

Comment:
```suggestion
        if (other.isNull && this.isNull) return true
        if (other.isNull || this.isNull) return false
```
Same goes for bag/list/sexp.

Reply:
Good catch. I think this would do the trick.

- true iff both null
- false iff only one is null
- continues if both non-null

```suggestion
if (this.isNull || other.isNull) return this.isNull == other.isNull
```